providers/dynamodb/shedlock-provider-dynamodb/src/main/java/net/javacrumbs/shedlock/provider/dynamodb/DynamoDBLockProvider.java

```diff
@@ -0,0 +1,238 @@
+import com.amazonaws.services.dynamodbv2.model.TableDescription;
+import net.javacrumbs.shedlock.core.LockConfiguration;
+import net.javacrumbs.shedlock.core.LockProvider;
+import net.javacrumbs.shedlock.core.SimpleLock;
+import net.javacrumbs.shedlock.support.Utils;
+
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.util.Optional;
+
+/**
+ * Distributed lock using DynamoDB.
+ * Depends on <code>aws-java-sdk-dynamodb</code>.
+ * <p>
+ * It uses a table with the following structure:
+ * <pre>
+ * {
+ *    "_id" : "lock name",
+ *    "lockUntil" : ISODate("2017-01-07T16:52:04.071Z"),
+ *    "lockedAt" : ISODate("2017-01-07T16:52:03.932Z"),
+ *    "lockedBy" : "host name"
+ * }
+ * </pre>
+ *
+ * <code>lockedAt</code> and <code>lockedBy</code> are just for troubleshooting
+ * and are not read by the code.
+ *
+ * <ol>
+ * <li>
+ * Attempts to insert a new lock record.
+ * </li>
+ * <li>
+ * We will try to update lock record using <code>filter _id == :name AND lock_until &lt;= :now</code>.
+ * </li>
+ * <li>
+ * If the update succeeded, we have the lock. If the update failed (condition check exception)
+ * somebody else holds the lock.
+ * </li>
+ * <li>
+ * When unlocking, <code>lock_until</code> is set to <i>now</i> or <i>lockAtLeastUntil</i> whichever is later.
+ * </li>
+ * </ol>
+ */
+public class DynamoDBLockProvider implements LockProvider {
+    static final String LOCK_UNTIL = "lockUntil";
+    static final String LOCKED_AT = "lockedAt";
+    static final String LOCKED_BY = "lockedBy";
+    static final String ID = "_id";
+
+    static final String DEFAULT_TABLE_NAME = "shedLock";
+
+    private static final String OBTAIN_LOCK_QUERY =
+            "set " + LOCK_UNTIL + " = :lockUntil, " + LOCKED_AT + " = :lockedAt, " + LOCKED_BY + " = :lockedBy";
+    private static final String OBTAIN_LOCK_CONDITION =
+            LOCK_UNTIL + " <= :lockedAt or attribute_not_exists(" + LOCK_UNTIL + ")";
+    private static final String RELEASE_LOCK_QUERY =
+            "set " + LOCK_UNTIL + " = :lockUntil";
+
+    /**
+     * A {@link DateTimeFormatter} like {@link DateTimeFormatter#ISO_INSTANT} with
+     * the exception that it always appends exactly three fractional digits (nano seconds).
+     * <p>
+     * This is required in order to guarantee natural sorting, which enables us to use
+     * <code>&lt;=</code> comparision in queries.
+     *
+     * <pre>
+     * 2018-12-07T12:30:37.000Z
+     * 2018-12-07T12:30:37.810Z
+     * 2018-12-07T12:30:37.819Z
+     * 2018-12-07T12:30:37.820Z
+     * </pre>
+     *
+     * When using variable fractional digit count as done in {@link DateTimeFormatter#ISO_INSTANT ISO_INSTANT}
+     * and {@link DateTimeFormatter#ISO_OFFSET_DATE_TIME ISO_OFFSET_DATE_TIME} the following sorting
+     * occurs:
+     *
+     * <pre>
+     * 2018-12-07T12:30:37.819Z
```

Comment:
Wow, I was not aware of that.

Reply:
I rebased my code on latest master and used `Utils.toIsoString`. Thank you.